.NET blog post (#1932)
pulumi/docs: content/blog/dotnet-preview/index.md

---
title: "Infrastructure as Code with .NET and Pulumi"
authors: ["sophia-parafina"]
tags: [". NET", "C#", "F#", "Visual Basic"]
meta_image: "microsoft_dotnet.png"
meta_desc: "Pulumi supports . NET languages for Infastructure as Code"
date: "2019-11-08"
---

We are excited to announce the preview of Pulumi .NET support. Pulumi lets you declare infrastructure using general purpose programming languages. Our goal is to automate applications at all layers and help developers and operators work better together by eliminating the silos that separate them.

> Millions of .NET developers can now use their favorite languages and open source ecosystems to build modern, cloud native applications. We've added support for C#, F#, and Visual Basic. Because the .NET  is available on Windows, Linux and macOS you have a choice of platforms to use. You can create cloud resources by writing **Microsoft .NET Core** programs to build and deploy AKS Clusters, Functions, Azure App Services, Virtual Machines, Cosmos DBs, KeyVaults, and much, much more.

Reviewer: ```suggestion
Millions of .NET developers can now use their favorite languages and open source ecosystems to build modern, cloud native applications. We've added support for C#, F#, and Visual Basic. Because the .NET  is available on Windows, Linux, and macOS, you have a choice of platforms to use. You can create cloud resources by writing **Microsoft .NET Core** programs to build and deploy AKS Clusters, Functions, Azure App Services, Virtual Machines, Cosmos DBs, KeyVaults, and much, much more.
```

## What Does Pulumi Enable?

> As a . NET developer you can:

Reviewer: There are still 6 instances of ". NET"

Author: fixed

> As a . NET developer you can:

Reviewer: ```suggestion
As a .NET developer, you can:
```

* Declare infrastructure in C#, F#, or Visual Basic NET.
* Code with favorite IDEs, including Visual Studio, Visual Studio Code, and Rider, with auto-completion and refactoring tools.
* Catch mistakes early on with red markers and build errors.
* Allow package managers to reference any NuGet library compatible with .NET Core 3.0.
* Use the Pulumi CLI or CI/CD integrations to perform deployments predictably and reliably.
* Maintain your infrastructure in a repository and version it
* Build modern, reliable, and scalable applications using a multitude of cloud architectures, including containers, serverless functions, and static websites.

You can do all of this without having to learn a DSL or script CLI commands. With Pulumi, developers and operators have a shared foundation for building modern applications.

## How Pulumi . NET Works

> Pulumi is declarative even though it uses general purpose programming languages. It does this by declaring a set of resources and the Pulumi engine orchestrates the CRUD operations to build and deploy infrastructure.

Reviewer: ```suggestion
Pulumi is declarative even though it uses general purpose programming languages. It does this by declaring a set of resources, and the Pulumi engine orchestrates the CRUD operations to build and deploy infrastructure.
```

> We use a .NET Core console application to declare infrastructure. Let's create an Azure App Service with a SQL database. The first step is make sure that your environment meets the prerequisites:

Reviewer: ```suggestion
We use a .NET Core console application to declare infrastructure. Let's create an Azure App Service with a SQL database. The first step is to make sure that your environment meets the prerequisites:
```

### Prerequisites

1. [Install Pulumi](https://www.pulumi.com/docs/get-started/install/)

1. [Install . NET Core 3.0+](https://dotnet.microsoft.com/download)

1. [A Microsoft Azure account](https://azure.microsoft.com/en-us/free/)

1. [Install Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli)

### Get the Examples

Pulumi provides a repository of examples to get you started. Clone the examples from GitHub.

   ``` bash
   $ git clone https://github.com/pulumi/examples/
   $ cd examples/
   ```

### Create a New Stack

Change the directory to `azure-cs-appservice` directory

   ``` bash
   $ cd azure-cs-appservice
   ```

> There are four files in the Pulumi project, the `Azure.Appservice.csproj` which sets the . NET SDK and the Pulumi packages for the project. `Pulumi.yaml` holds project parameters such as name and runtime. The `Pulumi.dev.yaml` file holds the database password. `SharedAccessSignature.cs` contains a helper function to generate storage blob access URLs with SAS tokens in them. It demonstrate that you can add custom code inside your Pulumi program. `Program.cs` is the main program for declaring resources.

Reviewer: ```suggestion
There are four files in the Pulumi project, the `Azure.Appservice.csproj`, which sets the .NET SDK and the Pulumi packages for the project. `Pulumi.yaml` holds project parameters such as name and runtime. The `Pulumi.dev.yaml` file holds the database password. `SharedAccessSignature.cs` contains a helper function to generate storage blob access URLs with SAS tokens in them. It demonstrates that you can add custom code inside your Pulumi program. `Program.cs` is the main program for declaring resources.
```

    ```

1. Run `pulumi up` to preview and deploy changes:

    ``` bash
    $ pulumi up
    Previewing changes:
    ...

    Performing changes:
    ...
    info: 10 changes performed:

        + 10 resources created

    Update duration: 1m14.59910109s
    ```

___
**NOTE:**

> If you are using a free Azure account, some regions do not allow creation of services if they are too busy. You might have to try a different [Azure location](https://azure.microsoft.com/en-us/global-infrastructure/locations/). You must tear down the stack resources to update the project's location information.
___

### Test Your Stack

Check the deployed website endpoint:

   ``` bash
   $ pulumi stack output endpoint
   https://azpulumi-as0ef47193.azurewebsites.net
   $ curl "$(pulumi stack output endpoint)"
   <html>
      <body>
         <h1>Greetings from Azure App Service!</h1>
      </body>
   </html>
   ```

From there, feel free to experiment. Simply making edits and running `pulumi up` will incrementally update your stack. Once finished, tear down your stack's resources by destroying and removing it:

   ``` bash
   $ pulumi destroy --yes
   $ pulumi stack rm --yes
   ```

## Recap

> In this post we've shown how to use Pulumi to deploy a modern application to Azure. The example code performs a number of tasks.

Reviewer: ```suggestion
In this post, we've shown how to use Pulumi to deploy a modern application to Azure. The example code performs a number of tasks.
```

   ```C#
   var container = new Container("zips", new ContainerArgs
   {
       StorageAccountName = storageAccount.Name,
       ContainerAccessType = "private",
   });

   var blob = new ZipBlob("zip", new ZipBlobArgs
   {
       StorageAccountName = storageAccount.Name,
       StorageContainerName = container.Name,
       Type = "block",
       Content = new FileArchive("wwwroot"),
   });

   var codeBlobUrl = SharedAccessSignature.SignedBlobReadUrl(blob, storageAccount);
   ```

Using the configuration we set with the Pulumi CLI, the program creates a SQLServer deployment.

   ```C#
   var config = new Config();
   var username = config.Get("sqlAdmin") ?? "pulumi";
   var password = config.RequireSecret("sqlPassword");
   var sqlServer = new SqlServer("sql", new SqlServerArgs
   {
       ResourceGroupName = resourceGroup.Name,
       AdministratorLogin = username,
       AdministratorLoginPassword = password,
       Version = "12.0",
   });

   var database = new Database("db", new DatabaseArgs
   {
       ResourceGroupName = resourceGroup.Name,
       ServerName = sqlServer.Name,
       RequestedServiceObjectiveName = "S0",
   });
   ```

> We deploy the app service which includes the website and the connection to the SQLServer instance.

Reviewer: ```suggestion
We deploy the app service, which includes the website and the connection to the SQLServer instance.
```

> We deploy the app service which includes the website and the connection to the SQLServer instance.

Reviewer: Don't we want to title-case service names, e.g., App Service?

Author: yes, accepted

---
title: "Infrastructure as Code with .NET and Pulumi"
date: 2019-11-13T11:23:04-06:00
draft: false

> meta_desc: "Pulumi supports .NET Core languages for Infrastructure as Code"

Reviewer: Maybe make it more... active? Like, `Pulumi launches the support`....

Author: accepted

> meta_image: "meta.png"

Reviewer 2: I'm seeing this render a bit funny (overflowing the bounding box).

![Screen Shot 2019-11-13 at 1 26 22 PM](https://user-images.githubusercontent.com/223467/68805761-4df3e500-0619-11ea-8460-d9fb71503aa0.png)

authors:
    - sophia-parafina
tags:

>     - ".Net Core"

Reviewer 3: ```suggestion
    - ".NET Core"
```

Author: ```suggestion
    - ".NET Core"
```

Author: fixed

>     - ".NET Core"

Reviewer 2: We used `.NET` for the post yesterday - lets reuse that here.

    - "C#"
    - "F#"
    - "Visual Basic"
---

> We are excited to announce the preview of Pulumi .NET support. Pulumi lets you declare infrastructure using general purpose programming languages. Our goal is to automate applications at all layers and help developers and operators work better together by eliminating the silos that separate them.

Reviewer: We already announced this support before?

Reviewer 2: Right - we posted https://www.pulumi.com/blog/pulumi-dotnet-core/ yesterday as an "announce".  I thought this post was focused on being more of a follow-up walkthrough of the experience using it?  Like taking https://www.pulumi.com/docs/intro/languages/dotnet/ and blowing it out into a few more pages of content?

> Millions of .NET developers can now use their favorite languages and open source ecosystems to build modern, cloud native applications. We've added support for C#, F#, and Visual Basic. Because .NET Core is available on Windows, Linux, and macOS, you have a choice of platforms to use. You can create cloud resources by writing Microsoft .NET Core programs to build and deploy cloud resources to a wide variety of clouds, including Azure, AWS, GCP and more.  On Azure, you can manage resources like AKS Clusters, Functions, Azure App Services, Virtual Machines, Cosmos DBs, KeyVaults, and much, much more.

Reviewer 2: I feel like we could lead in hear a bit better.  Perhaps reference yesterday's launch of support for .NET in Pulumi in preview? Perhaps note that in this post we'll dive deeper into how to get started with building infrastructure as code using .NET in Pulumi?

Author: With the release of [Pulumi for .NET preview](https://www.pulumi.com/blog/pulumi-dotnet-core/), we've open the doors to infrastructure as code to even more developers and operators. Millions of .NET developers can now use their favorite languages and open source ecosystems to build modern, cloud native applications. We've added support for C#, F#, and Visual Basic. Because .NET Core is available on Windows, Linux, and macOS, you have a choice of platforms to use.

You can create cloud resources by writing Microsoft .NET Core programs to build and deploy cloud resources to a wide variety of clouds, including Azure, AWS, GCP and more.  On Azure, you can manage resources like AKS Clusters, Functions, Azure App Services, Virtual Machines, Cosmos DBs, KeyVaults, and much, much more. Let's take a first look at Pulumi for .NET by deploying an application on Azure.

> Millions of .NET developers can now use their favorite languages and open source ecosystems to build modern, cloud native applications. We've added support for C#, F#, and Visual Basic. Because .NET Core is available on Windows, Linux, and macOS, you have a choice of platforms to use. You can create cloud resources by writing Microsoft .NET Core programs to build and deploy AKS Clusters, Functions, Azure App Services, Virtual Machines, Cosmos DBs, KeyVaults, and much, much more.

Reviewer 2: ```suggestion
Millions of .NET developers can now use their favorite languages and open source ecosystems to build modern, cloud native applications. We've added support for C#, F#, and Visual Basic .NET. Because .NET Core is available on Windows, Linux, and macOS, you have a choice of platforms to use. You can create cloud resources by writing Microsoft .NET Core programs to build and deploy cloud resources to a wide variety of clouds, including Azure, AWS, GCP and more.  On Azure, you can manage resources like AKS Clusters, Functions, Azure App Services, Virtual Machines, Cosmos DBs, KeyVaults, and much, much more.
```

## What Does Pulumi Enable?

> With Pulumi, .NET developers can declare infrastructure using C#, F#, or Visual Basic .NET. This results in increased productivity for developers because they can use the features of their IDEs such as auto-completion, error checking with red markers, build error messages, refactoring tools, and package managers. Developers can reference any NuGet library compatible with .NET Core 3.0.

Reviewer 2: ```suggestion
With Pulumi, .NET developers can create, deploy and manage cloud infrastructure using C#, F#, or Visual Basic .NET. This results in increased productivity for developers because they can use the features of their IDEs such as auto-completion, error checking with red markers, build error messages, refactoring tools, and package managers. Developers can reference any .NET libraries on NuGet compatible with .NET Core 3.0.
```

> Operators can use the Pulumi CLI or CI/CD integrations to maintain and version infrastructure in a repository resulting in predictable and reliable deployments. Regardless of your cloud architecture, whether it includes containers, serverless function or static websites, Pulumi lets you build modern, reliable, and scalable applications. Your pipeline is part of your product.

Reviewer 2: ```suggestion
Operators can use the `pulumi` CLI or CI/CD integrations to maintain and version infrastructure in a repository resulting in predictable and reliable deployments. Regardless of your cloud architecture, whether it includes containers, serverless function or static websites, Pulumi lets you build modern, reliable, and scalable applications. Your pipeline is part of your product.  Pulumi integrates with all of your favorite continuous delivery systems - including Azure DevOps, GitHub Actions and many more.
```

> ## How Pulumi .NET Works

Reviewer 2: Let's avoid saying "Pulumi .NET".  Let's use "Pulumi for .NET" or ".NET support in Pulumi" or similar.

> You can accomplish all of this without having to learn a DSL or script CLI commands. With Pulumi, developers and operators have a shared foundation for building modern applications.

Reviewer 2: ```suggestion
You can accomplish all of this without having to learn a JSON or YAML DSL or script low-level CLI commands. With Pulumi, developers and operators have a shared foundation for building modern applications.
```

## How Pulumi .NET Works

Pulumi is declarative even though it uses general purpose programming languages. It does this by declaring a set of resources, and the Pulumi engine orchestrates the CRUD operations to build and deploy infrastructure.

> A Pulumi project uses a .NET Core console application to declare infrastructure. To illustrate how this is accomplished, let's create an Azure App Service with an SQL database. The first step is to make sure that your environment meets the prerequisites:

Reviewer: Nit: three "declare" in three sentences, maybe reword?

Author: reworded

### Prerequisites

1. [Install Pulumi](https://www.pulumi.com/docs/get-started/install/)

> 1. [Install .NET Core 3.0+](https://dotnet.microsoft.com/download)

Reviewer 3: ```suggestion
1. [Install .NET Core SDK 3.0+](https://dotnet.microsoft.com/download)
```

Author: added

> 1. [A Microsoft Azure account](https://azure.microsoft.com/en-us/free/)

Reviewer: `Configure a Microsoft...`?

Reviewer 2: Agreed with @mikhailshilkov feedback.

1. [Install Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli)

### Get the Examples

Pulumi provides a repository of examples to get you started. Clone the examples from GitHub.

   ``` bash
   $ git clone https://github.com/pulumi/examples/
   $ cd examples/

>    ```

Reviewer 3: It's best not to indent these code blocks. Doing so adds extra indentation to the page that Hugo generates, which we don't want.

Author: indents removed

### Create a New Stack

Change the directory to `azure-cs-appservice` directory

   ``` bash
   $ cd azure-cs-appservice

>    ```

Reviewer 3: Unindent

> There are four files in the Pulumi project, the `Azure.Appservice.csproj` which sets the .NET SDK and the Pulumi packages for the project. `Pulumi.yaml` holds project parameters such as name and runtime. The `Pulumi.dev.yaml` file holds the database password. `SharedAccessSignature.cs` contains a helper function to generate storage blob access URLs with SAS tokens in them. It demonstrates that you can add custom code inside your Pulumi program. `Program.cs` is the main program for declaring resources.

Reviewer 2: ```suggestion
There are four files in the Pulumi project, the `Azure.Appservice.csproj` which sets the .NET SDK and the Pulumi packages for the project. `Pulumi.yaml` holds project parameters such as name and runtime. The `Pulumi.dev.yaml` file holds the database password to use for the `dev` environment as an encrypted secret. `SharedAccessSignature.cs` contains a helper function to generate storage blob access URLs with SAS tokens in them. It demonstrates that you can add custom code inside your Pulumi program. `Program.cs` is the main program for declaring resources.
```

1. Create a new stack:

    ``` bash
    $ pulumi stack init dev

>     ```

Reviewer 3: However, it is OK to indent these code blocks because they are part of the list of instructions, and we do want them indented along with the rest of the indented content for the list. Hugo renders these correctly.

1. Login to Azure CLI (you will be prompted to do this during deployment if you forget this step):

    ``` bash
    $ az login
    ```

1. Configure the location to deploy the resources to:

    ``` bash
    $ pulumi config set azure:location centralus
    ```

> 1. Define SQL Server password (make it complex enough to satisfy Azure policy):

Reviewer 2: Link to the Azure password policy?

Author: added this link:

https://docs.microsoft.com/en-us/azure/active-directory/authentication/concept-sspr-policy#password-policies-that-only-apply-to-cloud-user-accounts

    ``` bash

>     pulumi config set --secret sqlPassword <value>

Reviewer 3: ```suggestion
    $ pulumi config set --secret sqlPassword <value>
```

Author: fixed

    ```

> 1. Run `pulumi up` to preview and deploy changes:

Reviewer 2: Perhaps add a few more words here about what the "preview" and "update" are doing?

Author: Run `pulumi up`. Pulumi will display a preview of the updates to the stack and ask if you would like to deploy the by updating.

    ``` bash
    $ pulumi up
    Previewing changes:
    ...

    Performing changes:
    ...
    info: 10 changes performed:

        + 10 resources created

    Update duration: 1m14.59910109s
    ```

___
**NOTE:**

> If you are using a free Azure account, some regions do not allow creating services if they are too busy. You might have to try a different [Azure location](https://azure.microsoft.com/en-us/global-infrastructure/locations/). You must tear down the stack resources to update the project's location information.
___

### Test Your Stack

> Check the deployed website endpoint by querying the stack and using curl to retrieve the web page.

Reviewer 2: Perhaps add a note that `endpoint` is available because it was one of the stack outputs returned from the program.

Author: added

   ``` bash
   $ pulumi stack output endpoint
   https://azpulumi-as0ef47193.azurewebsites.net

   $ curl "$(pulumi stack output endpoint)"
   <html>
      <body>
         <h1>Greetings from Azure App Service!</h1>
      </body>
   </html>

>    ```

Reviewer 3: Unindent

From there, feel free to experiment. Simply making edits and running `pulumi up` will incrementally update your stack. Once finished, tear down your stack's resources by destroying and removing it:

   ``` bash
   $ pulumi destroy --yes
   $ pulumi stack rm --yes

>    ```

Reviewer 3: Unindent

## Summary

> In this post, we've shown how to use Pulumi to deploy a modern application to Azure. The example code performs a number of tasks.

Reviewer 2: I don't quite follow the "Summary" section being here - and then being followed by all the code details?  Is that intentional?  I see another section that looks like another Summary further below?

Author: retitled to: Code Walkthrough

and moved Summary to end.

First, we created a [resource group](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-overview#resource-groups), a [storage account](https://docs.microsoft.com/en-us/azure/storage/common/storage-redundancy-lrs), and an [application service plan](https://docs.microsoft.com/en-us/azure/app-service/overview-hosting-plans) for our application:

>    ```C#

Reviewer 3: We use `csharp` elsewhere, which works with the lang chooser. Best to just use `csharp` consistently throughout the website.

```suggestion
   ```csharp
```

   var resourceGroup = new ResourceGroup("appservice-rg"); 

   var storageAccount = new Account("sa", new AccountArgs
   {

       ResourceGroupName = resourceGroup.Name,
       AccountReplicationType = "LRS",
       AccountTier = "Standard",

   }); 

   var appServicePlan = new Plan("asp", new PlanArgs
   {

       ResourceGroupName = resourceGroup.Name,
       Kind = "App",
       Sku = new PlanSkuArgs
       {
           Tier = "Basic",
           Size = "B1",
       },

   });

>    ```

Reviewer 3: Unindent

The `wwwroot` directory containing the `index.html` page into is added to the storage account we created previously. The `SharedAccessSignature.cs` code is a helper function to generate storage blob access URLs with [SAS tokens](https://docs.microsoft.com/en-us/azure/storage/common/storage-sas-overview). This shows how you can add custom code inside your Pulumi program.  

>    ```C#

Reviewer 3: ```suggestion
   ```csharp
```

   var container = new Container("zips", new ContainerArgs
   {
       StorageAccountName = storageAccount.Name,
       ContainerAccessType = "private",
   });

   var blob = new ZipBlob("zip", new ZipBlobArgs
   {
       StorageAccountName = storageAccount.Name,
       StorageContainerName = container.Name,
       Type = "block",
       Content = new FileArchive("wwwroot"),
   });

   var codeBlobUrl = SharedAccessSignature.SignedBlobReadUrl(blob, storageAccount);

>    ```

Reviewer 3: Unindent

> Using the configuration we set with the Pulumi CLI, the program creates a SQLServer deployment.

Reviewer 3: ```suggestion
Using the configuration we set with the Pulumi CLI, the program creates a SQL Server deployment.
```

Author: fixed

>    ```C#

Reviewer 3: ```suggestion
   ```csharp
```

Author: fixed

   var config = new Config();
   var username = config.Get("sqlAdmin") ?? "pulumi";
   var password = config.RequireSecret("sqlPassword");
   var sqlServer = new SqlServer("sql", new SqlServerArgs
   {
       ResourceGroupName = resourceGroup.Name,
       AdministratorLogin = username,
       AdministratorLoginPassword = password,
       Version = "12.0",
   });

   var database = new Database("db", new DatabaseArgs
   {
       ResourceGroupName = resourceGroup.Name,
       ServerName = sqlServer.Name,
       RequestedServiceObjectiveName = "S0",
   });

>    ```

Reviewer 3: Unindent

We deploy the App Service which includes the website and the connection to the SQLServer instance.

>    ```C#

Reviewer 3: ```suggestion
   ```csharp
```

Author: fixed

   var app = new AppService("app", new AppServiceArgs
   {
       ResourceGroupName = resourceGroup.Name,
       AppServicePlanId = appServicePlan.Id,
       AppSettings =
       {
           { "WEBSITE_RUN_FROM_PACKAGE", codeBlobUrl },
       },
       ConnectionStrings =
           {
               new AppServiceConnectionStringsArgs
               {
                   Name = "db",
                   Type = "SQLAzure",
                   Value = Output.Tuple<string, string, string>(sqlServer.Name, database.Name, password).Apply(t =>
                   {
                       (string server, string database, string pwd) = t;
                       return $"Server= tcp:{server}.database.windows.net;initial catalog={database};userID={username};password={pwd};Min Pool Size=0;Max Pool Size=30;Persist Security Info=true;";
                   }),
               },
           },
       });

> ```

Reviewer 3: Unindent

Author: fixed

---
title: "Infrastructure as Code with .NET and Pulumi"
authors: ["sophia-parafina"]
tags: [". NET", "C#", "F#", "Visual Basic"]

> meta_image: "microsoft_logo.jpg"

Reviewer: This looks like a very old .NET logo... I'm not sure where to find the official materials, but they look somewhat like this one there days: https://docs.microsoft.com/en-us/dotnet/images/hub/dotnet-core.svg

Author: replaced with current logo

meta_desc: "Pulumi supports . NET languages for Infastructure as Code"
date: "2019-11-08"
---

We are excited to announce the preview of Pulumi . NET support. Pulumi lets you declare infrastructure using general purpose programming languages. Our goal is to automate applications at all layers and help developers and operators work better together by eliminating the silos that separate them.

> Millions of . NET developers can now use their favorite languages and open source ecosystems to build modern, cloud native applications. We've added support for C#, F# and Visual Basic. Because the . NET framework is available on Windows, Linux and macOS you have a choice of platforms to use. You can create cloud resources by writing **Microsoft . NET Core** programs to build and deploy AKS Clusters, Functions, Azure App Services, Virtual Machines, Cosmos DBs, KeyVaults, and much, much more.

Reviewer: Oxford comma for `, and Visual Basic`?

Author: I'm ok with the oxford coma

> Millions of . NET developers can now use their favorite languages and open source ecosystems to build modern, cloud native applications. We've added support for C#, F# and Visual Basic. Because the . NET framework is available on Windows, Linux and macOS you have a choice of platforms to use. You can create cloud resources by writing **Microsoft . NET Core** programs to build and deploy AKS Clusters, Functions, Azure App Services, Virtual Machines, Cosmos DBs, KeyVaults, and much, much more.

Reviewer: .NET framework, at this point, means the legacy framework which only runs on Windows. The modern cross-plat one is .NET Core.

Author: fixed the space between `. NET`, stupid markdown formatter

## What Does Pulumi Enable?

As a . NET developer you can:

* Declare infrastructure in C#, F#, or Visual Basic NET.
* Code with favorite IDEs, including Visual Studio, Visual Studio Code, and Rider, with auto-completion and refactoring tools.

> * Catch mistakes early on with red markers and build errors.

Reviewer: Isn't reusing the bits from the .NET blog post a problem? Also, SEO-wise -- might be considered "plagiarism"?

Author: rewrote this bit

> * Allow package managers to reference any NuGet library compatible with . NET Core 3.0.

Reviewer: Remove the space in `. NET`

Author: removed

* Use the Pulumi CLI or CI/CD integrations to perform deployments predictably and reliably.
* Maintain your infrastructure in a repository and version it
* Build modern, reliable, and scalable applications using a multitude of cloud architectures, including containers, serverless functions, and static websites.

You can do all of this without having to learn a DSL or script CLI commands. With Pulumi, developers and operators have a shared foundation for building modern applications.

## How Pulumi . NET Works

Pulumi is declarative even though it uses general purpose programming languages. It does this by declaring a set of resources and the Pulumi engine orchestrates the CRUD operations to build and deploy infrastructure.

> We use a . NET Core console application to declare infrastructure. Let's create an Azure App Service with a SQL database. The first step is make sure that your environment meets the prerequisites:

Reviewer: Again, should be ".NET" no spaces

Author: removed, I think that the markdown formatter is doing this.

### Prerequisites

1. [Install Pulumi](https://www.pulumi.com/docs/get-started/install/)

1. [Install . NET Core 3.0+](https://dotnet.microsoft.com/download)

1. [A Microsoft Azure account](https://azure.microsoft.com/en-us/free/)

1. [Install Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli)

### Get the Examples

Pulumi provides a repository of examples to get you started. Clone the examples from GitHub.

``` bash
git clone https://github.com/pulumi/examples/
cd examples/
```

### Create a New Stack

Change the directory to `azure-cs-appservice` directory

``` bash
$ cd azure-cs-appservice
```

> There are four files in the Pulumi project, the `Azure.Appservice.csproj` which sets the . NET SDK and the Pulumi packages for the project. `Pulumi.yaml` holds project parameters such as name, runtime and the database password. `SharedAccessSignature.cs` specifies the storage to hold code, which is the `wwwroot` directory in this project. `Program.cs` is the main program for declaring resources.

Reviewer: `Pulumi.yaml` doesn't hold the password - actual config values are in `Pulumi.stack.yaml`.

Author: corrected

> There are four files in the Pulumi project, the `Azure.Appservice.csproj` which sets the . NET SDK and the Pulumi packages for the project. `Pulumi.yaml` holds project parameters such as name, runtime and the database password. `SharedAccessSignature.cs` specifies the storage to hold code, which is the `wwwroot` directory in this project. `Program.cs` is the main program for declaring resources.

Reviewer: `SharedAccessSignature` contains a helper function to generate storage blob access URLs with SAS tokens in them. It's a demo that you can do any kind of custom code inside your Pulumi program. It doesn't have much to do with the `wwwroot` folder.

Author: rephrased it

1. Create a new stack:

    ``` bash
    $ pulumi stack init dev
    ```

1. Login to Azure CLI (you will be prompted to do this during deployment if you forget this step):

    ``` bash
    $ az login
    ```

1. Configure the location to deploy the resources to:

    ``` bash
    $ pulumi config set azure:location centralus
    ```

1. Define SQL Server password (make it complex enough to satisfy Azure policy):

    ``` bash
    pulumi config set --secret sqlPassword <value>
    ```

1. Run `pulumi up` to preview and deploy changes:

    ``` bash
    $ pulumi up
    Previewing changes:
    ...

    Performing changes:
    ...
    info: 10 changes performed:

        + 10 resources created

    Update duration: 1m14.59910109s
    ```

___
**NOTE:**

> If you are using a free Azure account, some regions do not allow creation of services if they are too busy. You might have to try a different [Azure location](https://azure.microsoft.com/en-us/global-infrastructure/locations/).
><br/><br/>
> You must tear down the stack resources to update the project's location information.
>
   ``` bash
   $ pulumi destroy --yes
   $ pulumi stack rm --yes

>    ```

Reviewer: I would be curious to see why this is needed... Maybe not worth adding this to the blog post, but rather to README?

Author: That was the problem I was having when I was switching between Azure locations. pulumi config would update the location unless I rm the stack
